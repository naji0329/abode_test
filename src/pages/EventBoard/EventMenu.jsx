import React, { useState } from "react";
import { IconButton, Menu, MenuItem } from "@mui/material";
import MoreVert from '@mui/icons-material/MoreVert';

export default ({ event, handleUpdate, handleDelete }) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const open = Boolean(anchorEl);

  const handleClick = (event) => { setAnchorEl(event.currentTarget); };
  const handleClose = () => { setAnchorEl(null); };
  const onUpdate = () => {
    handleUpdate(event);
    setAnchorEl(null);
  }
  const onDelete = () => {
    handleDelete(event);
    setAnchorEl(null);
  }
  
  return (
    <>
      <IconButton onClick={handleClick}><MoreVert /></IconButton>
      <Menu
        id="basic-menu"
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button',
        }}
      >
        <MenuItem onClick={onUpdate}>Update</MenuItem>
        <MenuItem onClick={onDelete}>Delete</MenuItem>
      </Menu>
    </>
  );
};