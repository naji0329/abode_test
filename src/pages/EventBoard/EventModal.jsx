import React, { useEffect, useState } from "react";
import { Button, Chip, ListItem, Modal, Paper, Stack, TextField } from "@mui/material";
import { DateTimePicker } from "@mui/x-date-pickers";
import toast from "react-hot-toast";
import styles from './index.module.css'
import dayjs from "dayjs";

export default ({ defaultValue, open, handleClose, handleConfirm }) => {
  const [invitees, setInvitees] = useState([]);
  const [invitee, setInvitee] = useState("");
  const [name, setName] = useState("");
  const [description, setDescription] = useState("");
  const [datetime, setDatetime] = useState(null);

  useEffect(() => {
    if(!defaultValue) return;

    setInvitees(defaultValue.invitees);
    setName(defaultValue.name);
    setDescription(defaultValue.description);
    setDatetime(dayjs(defaultValue.datetime));
  }, [defaultValue]);

  const onInviteeKeydown = (e) => {
    if(e.keyCode === 13) {
      if(invitees.find(data => data === invitee)) {
        toast.error('Email duplicated!');
        return;
      }
      const re = /\S+@\S+\.\S+/;
      if(!re.test(invitee)) {
        toast.error("Email is invalid");
        return;
      }
      setInvitees([...invitees, invitee]);
    }
  }

  const onDeleteInvitee = (data) => {
    setInvitees(invitees.filter(invitee => invitee != data));
  };

  return (
    <Modal
      open={open}
      onClose={handleClose}
    >
      <Stack className={styles['create-schedule-modal']} gap={2}>
        <TextField label="Event Name" sx={{ width: '90%' }} value={name} onChange={e => setName(e.target.value)} />
        <TextField label="Description" sx={{ width: '90%' }} value={description} onChange={e => setDescription(e.target.value)} />
        <DateTimePicker sx={{ width: '90%' }} value={datetime} onChange={v => setDatetime(v)} />
        <TextField sx={{ width: '90%' }}label="Add Invitees" value={invitee} onChange={e => setInvitee(e.target.value)} onKeyDown={onInviteeKeydown} />
        <Paper
          sx={{
            display: 'flex',
            justifyContent: 'center',
            flexWrap: 'wrap',
            listStyle: 'none',
            p: 0.5,
            m: 0
          }}
          component='ul'
        >
          { invitees.map((invitee, idx) => {
            return (
              <ListItem key={idx}>
                <Chip
                  label={invitee}
                  onDelete={() => onDeleteInvitee(invitee)}
                />
              </ListItem>
            );
          }) }
        </Paper>
        <Button variant="contained" onClick={() => handleConfirm({ description, name, invitees, datetime })}>{ defaultValue ? "Update" : "Create" }</Button>
      </Stack>
    </Modal>
  )
}