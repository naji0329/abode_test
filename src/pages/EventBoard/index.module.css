.create-schedule-modal {
  background-color: white;
  margin: 100px auto 0 auto;
  width: 500px;
  padding: 20px;
  display: flex;
  align-items: center;
}
.create-schedule-modal ul {
  border: 1px solid grey;
  width: 90%;
  min-height: 100px;
}
.create-schedule-modal li {
  width: fit-content;
  padding: 3px;
}
.schedule-container {
  min-width: 500px;
  margin-top: 100px;
}
.schedule-container button {
  width: fit-content;
}