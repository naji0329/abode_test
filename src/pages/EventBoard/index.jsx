import { Button, Container, Stack, Typography } from "@mui/material";
import React, { useEffect, useMemo, useState } from "react";
import styles from './index.module.css'
import { Toaster } from "react-hot-toast";
import { DataGrid } from "@mui/x-data-grid";
import { useDispatch, useSelector } from "react-redux";
import EventMenu from "./EventMenu";
import { createEvent, deleteEvent, listEvents, updateEvent } from "../../utils/apis";
import { signOut } from "../../slices/userSlice";
import { useNavigate } from "react-router-dom";
import EventModal from "./EventModal";

export default () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const [createModalOpen, setCreateModalOpen] = useState(false);
  const [updateModalOpen, setUpdateModalOpen] = useState(false);
  const [eventToUpdate, setEventToUpdate] = useState(null);

  const events = useSelector(state => state.event.events);
  const email = useSelector(state => state.user.email);

  const columns = useMemo(() => ([
    { field: 'name', headerName: 'Event Name', width: 100 },
    { field: 'description', headerName: 'Description', width: 200 },
    { field: 'datetime', headerName: 'Date Time', width: 200 },
    {
      field: 'invitees',
      headerName: 'Invitees',
      width: 200,
    },
    {
      field: '',
      headerName: '',
      width: 60,
      renderCell: (props) => {
        return <EventMenu 
          event={props.row} 
          handleDelete={handleDeleteEvent}
          handleUpdate={handleUpdateClick}
        />
      }
    }
  ]), []);

  useEffect(() => {
    dispatch(listEvents());
  }, []);

  const handleDeleteEvent = (event) => { dispatch(deleteEvent(event.id)); };
  const handleUpdateClick = (event) => {
    setEventToUpdate(event);
    setUpdateModalOpen(true);
  }

  const handleCreateEvent = ({ name, description, datetime, invitees }) => {
    dispatch(createEvent({ 
      name, 
      description, 
      datetime, 
      invitees 
    }, () => setCreateModalOpen(false)));
  };
  const handleUpdateEvent = (data) => {
    dispatch(updateEvent(
      eventToUpdate.id, 
      data, 
      () => setUpdateModalOpen(false)
    ));
  }

  const handleSignOut = () => {
    localStorage.removeItem('token');
    dispatch(signOut());
    navigate('/signup');
  }

  return (
    <Container>
      <Stack className={styles['schedule-container']} gap={1}>
        <Stack direction='row' justifyContent='space-between'>
          <Button variant="contained" onClick={() => setCreateModalOpen(true)}>Create Event</Button>
          <Stack direction='row' alignItems='center' gap={2}>
            <Typography>Welcome <i>{ email }</i></Typography>
            <Button variant="contained" color="error" onClick={handleSignOut}>Sign out</Button>
          </Stack>
        </Stack>
        <DataGrid
          rows={events}
          columns={columns}
          initialState={{
            pagination: {
              paginationModel: { page: 0, pageSize: 5 },
            },
          }}
          pageSizeOptions={[5, 10]}
          checkboxSelection
        />
      </Stack>
      <EventModal
        open={createModalOpen}
        handleClose={() => setCreateModalOpen(false)}
        handleConfirm={handleCreateEvent}
      />
      <EventModal
        open={updateModalOpen}
        defaultValue={eventToUpdate}
        handleClose={() => setUpdateModalOpen(false)}
        handleConfirm={handleUpdateEvent}
      />
      <Toaster />
    </Container>
  );
}