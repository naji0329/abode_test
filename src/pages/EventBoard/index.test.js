import React from 'react'
import { create, act } from 'react-test-renderer'
import EventBoard from './index'
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { thunk } from 'redux-thunk';

const mockUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUsedNavigate,
}));


describe('App test', () => {
  const initialState = { 
    event: { 
      events: [] 
    }, 
    user: { 
      email: '', 
      token: null 
    } 
  };
  const mockStore = configureStore([thunk]);
  let store;
  Object.defineProperty(window, 'scrollLeft', { value: {}, writable: true });


  it('should work', () => {
    let tree
    store = mockStore(initialState);

    act(() => {
      tree = create(
        <Provider store={store}>
          <EventBoard />
        </Provider>, 
        {
          createNodeMock: (node) => {
            return document.createElement(node.type)
          }
        }
      )
    })
    expect(tree).toMatchSnapshot()
  })
  afterAll(() => jest.resetModules())
})