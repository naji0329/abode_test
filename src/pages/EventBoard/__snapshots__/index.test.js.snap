// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test should work 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <div
    className="MuiStack-root css-hwnj0i-MuiStack-root"
  >
    <div
      className="MuiStack-root css-gmwslw-MuiStack-root"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Create Event
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        className="MuiStack-root css-1kq5d21-MuiStack-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Welcome 
          <i />
        </p>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Sign out
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      aria-colcount={6}
      aria-multiselectable={false}
      aria-rowcount={1}
      className="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-az8st9-MuiDataGrid-root"
      role="grid"
    >
      <div
        className="MuiDataGrid-main css-204u17-MuiDataGrid-main"
      >
        <div
          className="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
          role="presentation"
          style={
            {
              "lineHeight": "56px",
              "maxHeight": 56,
              "minHeight": 56,
            }
          }
        >
          <div
            className="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
            role="rowgroup"
          >
            <div
              aria-rowindex={1}
              className="css-yrdy0g-MuiDataGrid-columnHeaderRow"
              role="row"
            >
              <div
                aria-colindex={1}
                aria-sort="none"
                className="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor MuiDataGrid-columnHeaderCheckbox"
                data-field="__check__"
                onBlur={[Function]}
                onClick={[Function]}
                onDoubleClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                role="columnheader"
                style={
                  {
                    "height": 56,
                    "maxWidth": 50,
                    "minWidth": 50,
                    "width": 50,
                  }
                }
                tabIndex={0}
              >
                <div
                  className="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable={false}
                  role="presentation"
                >
                  <div
                    className="MuiDataGrid-columnHeaderTitleContainer"
                    role="presentation"
                  >
                    <div
                      className="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <span
                        className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                        onBlur={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={null}
                      >
                        <input
                          aria-label="Select all rows"
                          checked={false}
                          className="PrivateSwitchBase-input css-1m9pwf3"
                          data-indeterminate={false}
                          onChange={[Function]}
                          required={false}
                          tabIndex={-1}
                          type="checkbox"
                        />
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="CheckBoxOutlineBlankIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                        <span
                          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </span>
                    </div>
                  </div>
                </div>
                <div
                  className="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  onClick={[Function]}
                  onDoubleClick={[Function]}
                  onMouseDown={[Function]}
                  style={
                    {
                      "minHeight": 56,
                      "opacity": 1,
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
              <div
                aria-colindex={2}
                aria-label="Event Name"
                aria-sort="none"
                className="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor"
                data-field="name"
                onBlur={[Function]}
                onClick={[Function]}
                onDoubleClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                role="columnheader"
                style={
                  {
                    "height": 56,
                    "maxWidth": 100,
                    "minWidth": 100,
                    "width": 100,
                  }
                }
                tabIndex={-1}
              >
                <div
                  className="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable={false}
                  role="presentation"
                >
                  <div
                    className="MuiDataGrid-columnHeaderTitleContainer"
                    role="presentation"
                  >
                    <div
                      className="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <div
                        aria-label=""
                        aria-labelledby={null}
                        className="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                        data-mui-internal-clone-element={true}
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onMouseLeave={[Function]}
                        onMouseOver={[Function]}
                        onTouchEnd={[Function]}
                        onTouchStart={[Function]}
                      >
                        Event Name
                      </div>
                    </div>
                    <div
                      className="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                    >
                      <button
                        aria-label="Sort"
                        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                        disabled={false}
                        onBlur={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={-1}
                        title="Sort"
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                          data-testid="ArrowUpwardIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                          />
                        </svg>
                        <span
                          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    className="MuiDataGrid-menuIcon"
                  >
                    <button
                      aria-expanded={false}
                      aria-haspopup="menu"
                      aria-label="Menu"
                      aria-labelledby={null}
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                      data-mui-internal-clone-element={true}
                      disabled={false}
                      id=":r8:"
                      onBlur={[Function]}
                      onClick={[Function]}
                      onContextMenu={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseOver={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      type="button"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                        data-testid="TripleDotsVerticalIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
                <div
                  className="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  onClick={[Function]}
                  onDoubleClick={[Function]}
                  onMouseDown={[Function]}
                  style={
                    {
                      "minHeight": 56,
                      "opacity": 1,
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
              <div
                aria-colindex={3}
                aria-label="Description"
                aria-sort="none"
                className="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable MuiDataGrid-withBorderColor"
                data-field="description"
                onBlur={[Function]}
                onClick={[Function]}
                onDoubleClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                role="columnheader"
                style={
                  {
                    "height": 56,
                    "maxWidth": 200,
                    "minWidth": 200,
                    "width": 200,
                  }
                }
                tabIndex={-1}
              >
                <div
                  className="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable={false}
                  role="presentation"
                >
                  <div
                    className="MuiDataGrid-columnHeaderTitleContainer"
                    role="presentation"
                  >
                    <div
                      className="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <div
                        aria-label=""
                        aria-labelledby={null}
                        className="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                        data-mui-internal-clone-element={true}
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onMouseLeave={[Function]}
                        onMouseOver={[Function]}
                        onTouchEnd={[Function]}
                        onTouchStart={[Function]}
                      >
                        Description
                      </div>
                    </div>
                    <div
                      className="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                    >
                      <button
                        aria-label="Sort"
                        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                        disabled={false}
                        onBlur={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={-1}
                        title="Sort"
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                          data-testid="ArrowUpwardIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                          />
                        </svg>
                        <span
                          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                        />
                      </button>
                    </div>
                  </div>
                  <div
                    className="MuiDataGrid-menuIcon"
                  >
                    <button
                      aria-expanded={false}
                      aria-haspopup="menu"
                      aria-label="Menu"
                      aria-labelledby={null}
                      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                      data-mui-internal-clone-element={true}
                      disabled={false}
                      id=":re:"
                      onBlur={[Function]}
                      onClick={[Function]}
                      onContextMenu={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseOver={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      type="button"
                    >
                      <svg
                        aria-hidden={true}
                        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                        data-testid="TripleDotsVerticalIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
                <div
                  className="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  onClick={[Function]}
                  onDoubleClick={[Function]}
                  onMouseDown={[Function]}
                  style={
                    {
                      "minHeight": 56,
                      "opacity": 1,
                    }
                  }
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="MuiDataGrid-virtualScroller css-qvtrhg-MuiDataGrid-virtualScroller"
          onScroll={[Function]}
          onTouchMove={[Function]}
          onWheel={[Function]}
          role="presentation"
          style={
            {
              "overflowX": "hidden",
            }
          }
        >
          <div
            className="MuiDataGrid-overlayWrapper css-ql19fo-MuiDataGrid-overlayWrapper"
          >
            <div
              className="MuiDataGrid-overlayWrapperInner css-1akuw9y-MuiDataGrid-overlayWrapperInner"
              style={
                {
                  "height": 0,
                  "width": 0,
                }
              }
            >
              <div
                className="MuiDataGrid-overlay css-1w53k9d-MuiDataGrid-overlay"
              >
                No rows
              </div>
            </div>
          </div>
          <div
            className="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
            role="presentation"
            style={
              {
                "height": 1,
                "minHeight": "auto",
                "width": "auto",
              }
            }
          >
            <div
              className="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
              role="rowgroup"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiDataGrid-footerContainer MuiDataGrid-withBorderColor css-wop1k0-MuiDataGrid-footerContainer"
      >
        <div />
        <div
          className="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
        >
          <div
            className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
          >
            <div
              className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
            />
            <p
              className="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
              id=":r1:"
            >
              Rows per page:
            </p>
            <div
              className="MuiInputBase-root MuiInputBase-colorPrimary MuiTablePagination-input css-16c50h-MuiInputBase-root-MuiTablePagination-select"
              onClick={[Function]}
            >
              <div
                aria-controls=":r2:"
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby=":r1: :r0:"
                className="MuiSelect-select MuiTablePagination-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
                id=":r0:"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                role="combobox"
                tabIndex={0}
              >
                5
              </div>
              <input
                aria-hidden={true}
                aria-invalid={false}
                className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                disabled={false}
                onAnimationStart={[Function]}
                onChange={[Function]}
                tabIndex={-1}
                value={5}
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiTablePagination-selectIcon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
            <p
              className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
            >
              0–0 of 0
            </p>
            <div
              className="MuiTablePagination-actions"
            >
              <button
                aria-label="Go to previous page"
                className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Go to previous page"
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowLeftIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                  />
                </svg>
              </button>
              <button
                aria-label="Go to next page"
                className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                title="Go to next page"
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      {
        "bottom": 16,
        "left": 16,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 16,
        "top": 16,
        "zIndex": 9999,
      }
    }
  />
</div>
`;
